File: archiso/airootfs/usr/share/churros/preferences/pages/connectivity.py
import threading
import gi

gi.require_version("Gtk", "4.0")

from gi.repository import GLib, Gtk

from widgets.page import Page
from widgets.group import Group
from widgets.row import Row
from widgets.switch_row import SwitchRow

from services.connectivity import ConnectivityService


class ConnectivityPage(Page):

    def __init__(self, navigator):

        super().__init__(
            navigator,
            "Conectividad",
            "Wi-Fi y Bluetooth"
        )

        self.wifi_group = Group("Wi-Fi")
        self.bluetooth_group = Group("Bluetooth")

        self.add(self.wifi_group)
        self.add(self.bluetooth_group)

        self.wifi_group.add(Row(title="Cargando..."))
        self.bluetooth_group.add(Row(title="Cargando..."))

        threading.Thread(target=self.load, daemon=True).start()

    def load(self):

        data = {
            "wifi_available": ConnectivityService.wifi_available(),
            "wifi_enabled": ConnectivityService.wifi_enabled(),
            "current_network": ConnectivityService.current_network(),
            "wifi_networks": ConnectivityService.wifi_networks_full(),
            "bluetooth_available": ConnectivityService.bluetooth_available(),
            "bluetooth_enabled": ConnectivityService.bluetooth_enabled(),
            "bluetooth_devices": ConnectivityService.bluetooth_devices(),
        }

        GLib.idle_add(self.populate, data)

    def populate(self, data):

        self.wifi_group.clear()

        if not data["wifi_available"]:
            self.wifi_group.add(
                Row(title="No se encontró un adaptador Wi-Fi")
            )
        else:
            self.wifi_group.add(
                SwitchRow(
                    title="Activar Wi-Fi",
                    active=data["wifi_enabled"],
                    callback=self.on_wifi
                )
            )

            current = data["current_network"]

            if current:
                current_row = Row(
                    title="Red actual",
                    subtitle=current,
                    callback=lambda *_: self._prompt_forget(current)
                )
                self.wifi_group.add(current_row)

            networks = data["wifi_networks"]

            if networks:
                for net in networks:
                    self.wifi_group.add(self._make_network_row(net, current))
            else:
                self.wifi_group.add(
                    Row(title="No se encontraron redes")
                )

            self.wifi_group.add(
                Row(
                    title="Recargar redes",
                    subtitle="Forzar un nuevo escaneo",
                    callback=lambda *_: GLib.timeout_add(250, self.reload)
                )
            )

        #
        # Bluetooth
        #

        self.bluetooth_group.clear()

        if not data["bluetooth_available"]:
            self.bluetooth_group.add(
                Row(title="No se encontró un adaptador Bluetooth")
            )
        else:
            self.bluetooth_group.add(
                SwitchRow(
                    title="Activar Bluetooth",
                    active=data["bluetooth_enabled"],
                    callback=self.on_bluetooth
                )
            )

            if data["bluetooth_devices"]:
                for device in data["bluetooth_devices"]:
                    self.bluetooth_group.add(
                        Row(
                            title=device["name"],
                            subtitle=device["mac"]
                        )
                    )
            else:
                self.bluetooth_group.add(
                    Row(title="No hay dispositivos")
                )

        return False

    def _make_network_row(self, net, current):

        ssid = net.get("ssid", "?")
        signal = net.get("signal", 0)
        security = net.get("security", "")
        connected = net.get("connected", False) or ssid == current
        saved = net.get("saved", False)

        subtitle_parts = ["Señal: {}%".format(signal)]

        if security:
            subtitle_parts.append(security)

        if connected:
            subtitle_parts.append("conectado")
        elif saved:
            subtitle_parts.append("guardada")

        subtitle = " · ".join(subtitle_parts)

        row = Row(
            title=ssid,
            subtitle=subtitle,
            callback=lambda *_, s=ssid, sec=security: self._on_network_clicked(s, sec)
        )

        gesture = Gtk.GestureClick()
        gesture.set_button(3)
        gesture.connect(
            "pressed",
            lambda *_, s=ssid: self._prompt_forget(s)
        )
        row.add_controller(gesture)

        return row

    def _on_network_clicked(self, ssid, security):

        if not security:
            self._do_connect(ssid, None)
            return

        self._prompt_password(ssid)

    def _prompt_password(self, ssid):

        dialog = Gtk.AlertDialog()
        dialog.set_heading("Conectar a {}".format(ssid))
        dialog.set_modal(True)
        dialog.set_buttons(["Cancelar", "Conectar"])

        label = Gtk.Label(
            label="Introduce la contraseña de la red Wi-Fi:"
        )
        label.set_margin_top(14)
        label.set_margin_start(14)
        label.set_margin_end(14)

        entry = Gtk.PasswordEntry()
        entry.set_show_peek_icon(True)
        entry.set_margin_start(14)
        entry.set_margin_end(14)
        entry.set_margin_bottom(14)

        box = Gtk.Box(
            orientation=Gtk.Orientation.VERTICAL,
            spacing=12
        )
        box.append(label)
        box.append(entry)

        dialog.set_extra_child(box)

        entry.connect("activate", lambda *_: dialog.choose_finish(dialog.choose(self.get_root())))

        dialog.choose(
            self.get_root(),
            None,
            lambda d, r: self._on_password_response(d, r, ssid, entry)
        )

    def _on_password_response(self, dialog, result, ssid, entry):

        try:

            response = dialog.choose_finish(result)

        except Exception:

            return

        if response != 1:
            return

        password = entry.get_text() or ""

        if not password:
            self._show_toast("Se requiere contraseña para esa red.")
            return

        self._do_connect(ssid, password)

    def _do_connect(self, ssid, password):

        def work():

            ok, err = ConnectivityService.wifi_connect(ssid, password)

            if ok:
                GLib.idle_add(self.reload)
            else:
                GLib.idle_add(self._show_toast, "No se pudo conectar: " + (err or "error desconocido"))

        threading.Thread(target=work, daemon=True).start()

    def _prompt_forget(self, ssid):

        dialog = Gtk.AlertDialog()
        dialog.set_heading("Olvidar red")
        dialog.set_message("¿Olvidar la configuración de la red '{}'?".format(ssid))
        dialog.set_modal(True)
        dialog.set_buttons(["Cancelar", "Olvidar"])

        dialog.choose(
            self.get_root(),
            None,
            lambda d, r: self._on_forget_response(d, r, ssid)
        )

    def _on_forget_response(self, dialog, result, ssid):

        try:

            response = dialog.choose_finish(result)

        except Exception:

            return

        if response != 1:
            return

        def work():

            ConnectivityService.wifi_forget(ssid)
            GLib.idle_add(self.reload)

        threading.Thread(target=work, daemon=True).start()

    def _show_toast(self, message):

        try:

            root = self.get_root()

            if root is not None and hasattr(root, "toast"):

                root.toast.add_notification(message)

                return

        except Exception:

            pass

        print("[connectivity]", message)

    def reload(self):

        threading.Thread(target=self.load, daemon=True).start()

        return False

    def on_wifi(self, active):

        ConnectivityService.set_wifi(active)

        if active:
            GLib.timeout_add(1000, self.reload)
        else:
            GLib.timeout_add(250, self.reload)

    def on_bluetooth(self, active):

        ConnectivityService.set_bluetooth(active)

File: rust/preferences/src/pages/connectivity.rs
// ==========================================
// ConnectivityPage — Wi-Fi y Bluetooth
// (equivalente a pages/connectivity.py)
// ==========================================

use gtk::prelude::*;

use std::cell::RefCell;
use std::rc::Rc;

use crate::services::connectivity::ConnectivityService;
use crate::widgets::group::Group;
use crate::widgets::page::Page;
use crate::widgets::row::Row;
use crate::widgets::switch_row::SwitchRow;

// Datos Send (sin Rc ni widgets) para cargar en thread
#[derive(Default)]
struct WifiData {
    available: bool,
    enabled: bool,
    current: Option<String>,
    networks: Vec<(String, i64, String, bool, bool)>, // ssid, signal, security, connected, saved
}

#[derive(Default)]
struct BtData {
    available: bool,
    enabled: bool,
    devices: Vec<(String, String)>, // name, mac
}

struct ConnectivityData {
    wifi: WifiData,
    bluetooth: BtData,
}

fn load_data() -> ConnectivityData {
    ConnectivityData {
        wifi: WifiData {
            available: ConnectivityService::wifi_available(),
            enabled: ConnectivityService::wifi_enabled(),
            current: ConnectivityService::current_network(),
            networks: ConnectivityService::wifi_networks_full()
                .into_iter()
                .map(|n| (n.ssid, n.signal, n.security, n.connected, n.saved))
                .collect(),
        },
        bluetooth: BtData {
            available: ConnectivityService::bluetooth_available(),
            enabled: ConnectivityService::bluetooth_enabled(),
            devices: ConnectivityService::bluetooth_devices()
                .into_iter()
                .map(|d| (d.name, d.mac))
                .collect(),
        },
    }
}

pub fn build(navigator: gtk::Stack) -> Page {
    let page = Page::new(
        Some(navigator),
        "Conectividad",
        Some("Wi-Fi y Bluetooth"),
        None,
    );

    let wifi_group = Rc::new(RefCell::new(Group::new("Wi-Fi")));
    let bluetooth_group = Rc::new(RefCell::new(Group::new("Bluetooth")));

    wifi_group
        .borrow_mut()
        .add(&Row::new("Cargando...", None, None, None, None, None));
    bluetooth_group
        .borrow_mut()
        .add(&Row::new("Cargando...", None, None, None, None, None));

    page.add(wifi_group.borrow().widget());
    page.add(bluetooth_group.borrow().widget());

    // Cargar datos en thread (nmcli/bluetoothctl pueden tardar).
    // Los Rc<RefCell<Group>> NO son Send: el thread solo produce datos y
    // los envía por canal; el hilo principal los consume en un timeout.
    let (tx, rx) = std::sync::mpsc::channel::<ConnectivityData>();
    std::thread::spawn(move || {
        let _ = tx.send(load_data());
    });

    let wifi_group_ui = Rc::clone(&wifi_group);
    let bluetooth_group_ui = Rc::clone(&bluetooth_group);
    glib::timeout_add_local(std::time::Duration::from_millis(100), move || {
        match rx.try_recv() {
            Ok(data) => {
                populate(&wifi_group_ui, &bluetooth_group_ui, data);
                glib::ControlFlow::Break
            }
            Err(_) => glib::ControlFlow::Continue,
        }
    });

    page
}

fn populate(
    wifi_group: &Rc<RefCell<Group>>,
    bluetooth_group: &Rc<RefCell<Group>>,
    data: ConnectivityData,
) {
    // ============ Wi-Fi ============
    wifi_group.borrow_mut().clear();

    if !data.wifi.available {
        wifi_group.borrow_mut().add(&Row::new(
            "No se encontró un adaptador Wi-Fi",
            None,
            None,
            None,
            None,
            None,
        ));
    } else {
        let wifi_group_ui = Rc::clone(wifi_group);
        wifi_group.borrow_mut().add(&SwitchRow::new(
            "Activar Wi-Fi",
            None,
            None,
            data.wifi.enabled,
            Some(Box::new(move |active| {
                ConnectivityService::set_wifi(active);
                let _ = &wifi_group_ui;
            })),
        ));

        if let Some(current) = &data.wifi.current {
            let current_owned = current.clone();
            wifi_group.borrow_mut().add(&Row::new(
                "Red actual",
                Some(&current_owned),
                None,
                None,
                None,
                None,
            ));
        }

        if data.wifi.networks.is_empty() {
            wifi_group.borrow_mut().add(&Row::new(
                "No se encontraron redes",
                None,
                None,
                None,
                None,
                None,
            ));
        } else {
            for (ssid, signal, security, connected, saved) in &data.wifi.networks {
                let mut parts = vec![format!("Señal: {signal}%")];
                if !security.is_empty() {
                    parts.push(security.clone());
                }
                if *connected {
                    parts.push("conectado".to_string());
                } else if *saved {
                    parts.push("guardada".to_string());
                }
                let subtitle = parts.join(" · ");

                let ssid_owned = ssid.clone();
                let security_owned = security.clone();
                wifi_group.borrow_mut().add(&Row::new(
                    ssid,
                    Some(&subtitle),
                    None,
                    None,
                    None,
                    Some(Box::new(move |_btn| {
                        if security_owned.is_empty() {
                            // Red abierta: conectar directo en thread
                            let ssid_for_thread = ssid_owned.clone();
                            std::thread::spawn(move || {
                                let (ok, err) =
                                    ConnectivityService::wifi_connect(&ssid_for_thread, None);
                                let _ = (ok, err);
                            });
                        } else {
                            // TODO: diálogo de contraseña (AlertDialog con PasswordEntry)
                            eprintln!("[connectivity] password dialog pendiente de portar");
                        }
                    })),
                ));
            }
        }

        wifi_group.borrow_mut().add(&Row::new(
            "Recargar redes",
            Some("Forzar un nuevo escaneo"),
            None,
            None,
            None,
            None,
        ));
    }

    // ============ Bluetooth ============
    bluetooth_group.borrow_mut().clear();

    if !data.bluetooth.available {
        bluetooth_group.borrow_mut().add(&Row::new(
            "No se encontró un adaptador Bluetooth",
            None,
            None,
            None,
            None,
            None,
        ));
    } else {
        let bluetooth_group_ui = Rc::clone(bluetooth_group);
        bluetooth_group.borrow_mut().add(&SwitchRow::new(
            "Activar Bluetooth",
            None,
            None,
            data.bluetooth.enabled,
            Some(Box::new(move |active| {
                ConnectivityService::set_bluetooth(active);
                let _ = &bluetooth_group_ui;
            })),
        ));

        if data.bluetooth.devices.is_empty() {
            bluetooth_group.borrow_mut().add(&Row::new(
                "No hay dispositivos",
                None,
                None,
                None,
                None,
                None,
            ));
        } else {
            for (name, mac) in &data.bluetooth.devices {
                bluetooth_group
                    .borrow_mut()
                    .add(&Row::new(name, Some(mac), None, None, None, None));
            }
        }
    }
}
